### PR TITLE
refactor: improve performance for incremental clustering and add type hints

### PR DESCRIPTION
**Pull request by anujsinha3**
#26 #29 #32
- [x] performance improvement by user-based indexing
- [x] add Python type hints
- [x] add folder structure for updated workflows (to allow UW ThinkLab team to start testing)
- [x] use logger instead of print statement

### CODE REVIEW

File: src/mawpy/steps/incremental_clustering.py [hunk 1]
```diff
@@ -69,17 +69,15 @@ def _merge_stays(stay_to_update, updated_stay, df_by_user, group_avgs, group_avg
     new_avg = merged_values.apply(_mean_ignore_minus_ones).fillna(-1)
 
     group_avgs.loc[group_avgs[STAY] == updated_stay, [STAY_LAT, STAY_LONG]] = new_avg.values
-    df_by_user.loc[df_by_user[STAY] == updated_stay, STAY_LAT] = new_avg.values[0]
-    df_by_user.loc[df_by_user[STAY] == updated_stay, STAY_LONG] = new_avg.values[1]
+    df_by_user.loc[df_by_user[STAY] == updated_stay, [STAY_LAT, STAY_LONG]] = new_avg.values
```
Comment (lsetiawan) [inline src/mawpy/steps/incremental_clustering.py:72]: `[STAY_LAT, STAY_LONG]` is repeated multiple times, can you create a new variable called `STAY_LAT_LONG`?

File: src/mawpy/steps/incremental_clustering.py [hunk 2]
```diff
@@ -288,8 +284,7 @@ def _get_locations_to_cluster_center_map(clusters_list):
     return locations_to_cluster_center_map
 
 
-def _run_for_user(df_by_user, spat_constr, dur_constr=None):
-
+def _run_for_user(df_by_user: pd.DataFrame, spat_constr: float, dur_constr: Optional[float] = None) -> pd.DataFrame:
```
Comment (lsetiawan) [inline src/mawpy/steps/incremental_clustering.py:287]:
```suggestion
def _run_for_user(df_by_user: pd.DataFrame, spat_constr: float, dur_constr: float | None = None) -> pd.DataFrame:
```

The new way of doing Optional is actually just do `type | None`.

File: src/mawpy/workflows/workflow1.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    duration_constraint2,
+):
+    incremental_clustering(input_file, output_file, spatial_constraint, duration_constraint1)
+
+
+if __name__ == "__main__":
+
+    multiprocessing.freeze_support()
+    st = datetime.datetime.now()
+    workflow1("/Users/anujsinha/MAWpy/src/mawpy/input_file_old.csv", "output_file.csv", 1.0, 0, 300)
```
Comment (lsetiawan) [inline src/mawpy/workflows/workflow1.py:23]: Can you update this to use [`argparse`](https://docs.python.org/3/library/argparse.html) so that you can specify the input file and output file on the command line?

**Review by lsetiawan — CHANGES_REQUESTED** [review]
A couple of tweaks needed 😄

File: src/mawpy/workflows/workflow1.py (hunk 3 above)
Comment (anujsinha3) [reply src/mawpy/workflows/workflow1.py]: added argparse

File: src/mawpy/steps/incremental_clustering.py (hunk 2 above)
Comment (anujsinha3) [reply src/mawpy/steps/incremental_clustering.py]: updated.

File: src/mawpy/steps/incremental_clustering.py (hunk 1 above)
Comment (anujsinha3) [reply src/mawpy/steps/incremental_clustering.py]: updated.

File: src/mawpy/constants.py [hunk 4]
```diff
@@ -1,13 +1,13 @@
 USER_ID = 'user_ID'
 STAY = 'stay'
 STAY_DUR = 'stay_dur'
-STAY_LAT = '             stay_lat' #TODO: currently sample csv shared (this is one of the examples with additional spaces
+STAY_LAT = 'stay_lat'
```
Comment (carlosgjs) [inline src/mawpy/constants.py:4]: Do we need a new/updated test file with this change?

File: src/mawpy/steps/incremental_clustering.py [hunk 5]
```diff
@@ -53,33 +52,32 @@ def _get_cluster_center(row, mapping, dur_constr):
...
     group_avgs.loc[group_avgs_index_to_update, STAY] = updated_stay
-    group_avgs.loc[group_avgs_index_to_update, STAY_LAT] = new_avg.values[0]
-    group_avgs.loc[group_avgs_index_to_update, STAY_LONG] = new_avg.values[1]
+    group_avgs.loc[group_avgs_index_to_update, STAY_LAT_LONG] = new_avg.values
 
     return df_by_user
 
 
-def _get_combined_stay(df_by_user, threshold=1):
+def _get_combined_stay(df_by_user: pd.DataFrame, threshold: float = 0.2) -> pd.DataFrame:
```
Comment (carlosgjs) [inline src/mawpy/steps/incremental_clustering.py:80]: Just checking if the change in the threshold value is intentional?

File: src/mawpy/steps/incremental_clustering.py [hunk 6]
```diff
@@ -215,7 +211,7 @@ def _k_means_cluster_lloyd(cluster_list):
     return cluster_list_new
 
 
-def _get_clusters(locations_for_clustering, spat_constr):
+def _get_clusters(locations_for_clustering: list[tuple[str, str]], spat_constr: float) -> list[Cluster]:
```
Comment (carlosgjs) [inline src/mawpy/steps/incremental_clustering.py:214]: Confirming we don't need to support python <3.9? I believe using the lower case hints is supported on >=3.9.

File: src/mawpy/steps/incremental_clustering.py [hunk 7]
```diff
@@ -345,41 +338,40 @@ def _run(args):
...
-def _divide_chunks(user_id_list, n):
-    user_id_chunks = []
-    for i in range(0, len(user_id_list), n):  # looping till length usernamelist
-        user_id_chunks.append(user_id_list[i: i + n])
-    return user_id_chunks
-
-
-def incremental_clustering(input_file, output_file, spatial_constraint, dur_constraint):
+def incremental_clustering(input_file: str, output_file: str, spatial_constraint: float, dur_constraint: float) -> None:
+    # this_lock = Lock()  # thread locker
```
Comment (carlosgjs) [inline src/mawpy/steps/incremental_clustering.py:342]: looks like the lock was not used or is not needed? If so we can remove the commented code. Can we also remove `init(...)` in line 32?

File: src/mawpy/workflows/workflow1.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    duration_constraint2,
+):
+    incremental_clustering(input_file, output_file, spatial_constraint, duration_constraint1)
+
+
+if __name__ == "__main__":
+    args = parser.parse_args()
+    multiprocessing.freeze_support()
+    st = datetime.datetime.now()
+    workflow1(args.input_file, "output_file.csv", 1.0, 0, 300)
```
Comment (carlosgjs) [inline src/mawpy/workflows/workflow1.py:32]: This could be part of another PR, but rather than using a hard-coded temp file name `output_file.csv` we should use the python temp file APIs. This will avoid conflicts with left over files in case of failed runs and will also allow for mutliple concurrent runs not conflicting with each other. Can you log an issue to address this across all workflows?

**Review by carlosgjs — CHANGES_REQUESTED** [review]
Some comments/questions

File: src/mawpy/constants.py (hunk 4 above)
Comment (anujsinha3) [reply src/mawpy/constants.py:4]: The last input csv file that Anurag shared is compatible with this.

File: src/mawpy/steps/incremental_clustering.py (hunk 6 above)
Comment (anujsinha3) [reply src/mawpy/steps/incremental_clustering.py:214]: The package has been designed keeping this in mind:
https://scientific-python.org/specs/spec-0000/

Anyhow, we can confirm this with ThinkLab team as well.

File: src/mawpy/steps/incremental_clustering.py (hunk 5 above)
Comment (anujsinha3) [reply src/mawpy/steps/incremental_clustering.py]: I asked UW ThinkLab to confirm the value in the last meeting, I have added a TODO as well now.

File: src/mawpy/steps/incremental_clustering.py (hunk 7 above)
Comment (anujsinha3) [reply src/mawpy/steps/incremental_clustering.py]: done

File: src/mawpy/workflows/workflow1.py (hunk 8 above)
Comment (anujsinha3) [reply src/mawpy/workflows/workflow1.py:32]: https://github.com/uw-ssec/MAWpy/issues/34

File: src/mawpy/steps/incremental_clustering.py (hunk 6 above)
Comment (carlosgjs) [reply src/mawpy/steps/incremental_clustering.py:210]: Ok, I see the pyproject.toml has this:
```
requires-python = ">=3.11"
```

**Review by carlosgjs — APPROVED** [review]
Looks good to go!

**Review by lsetiawan — APPROVED** [review]
